haystack/nodes/preprocessor/preprocessor_new.py

```diff
@@ -0,0 +1,475 @@
+from typing import List, Optional, Union
+
+from pathlib import Path
+import logging
+import warnings
+from tqdm.auto import tqdm
+from transformers import PreTrainedTokenizer
+
+from haystack.nodes.base import BaseComponent
+from haystack.schema import Document
+from haystack.nodes.preprocessor.splitter import DocumentSplitter, SplitBy
+from haystack.nodes.preprocessor.cleaner import DocumentCleaner
+
+
+logger = logging.getLogger(__name__)
+
+
+REGEX_METACHARS = r".^$*+?{}[]\|()"
+
+
+# NOTE: Is this class even useful? Can we do better than a hardcoded pair of splitter and cleaner?
+class NewPreProcessor(BaseComponent):
```

Comment:
First:  I guarantee this is a temporary name and it will not be merged as it is.

On this class specifically I have asked @robpasternak (which is reviewing from Sols' perspective) how useful this class really is and how it will be used, now that there are more granular classes. For the name I had in mind `DocumentPreprocessor` (to highlight that this is an indexing node), but I first planned to hear feedback about how it is used.

Regarding the deprecation, I agreed with @robpasternak that it will be at minimum the typical 2 releases cycle, but most likely stretch out as much as the old `PreProcessor` will be needed, up to Haystack 2.0 if so needed. Having two different names, nothing special about the deprecation routine: `PreProcessor` will have a deprecation warning asking users to migrate as soon as possible and then be dropped when it's safe to do so. If we keep the two classes for longer than two versions, we might even add a loud, `ERROR` level "final warning" in the last release that supports it and, in the first release that drops support, keep the class as an empty shell that raises an exception.  Still defining the details though.

Reply:
My intuition here is that `NewPreProcessor` will get plenty of usage, especially (but far from exclusively) by clients. For example, while I'm not a decisionmaker here my guess is that when the dC pipeline template YAMLs are eventually updated, the `NewPreProcessor` will be what's in the YAML (for exactly the reasons @tholor gives). The more granular subclasses will no doubt get used as well, but my guess is we'd use `NewPreProcessor` by default and then switch to the more fine-grained options at later stages in development or in cases where something's not working like we hoped.